make-disk-image: documentation, UEFI variables recording, improved determinism (#207038)
NixOS/nixpkgs: doc/builders/images/makediskimage.section.md

# `<nixpkgs/nixos/lib/make-disk-image.nix>` {#sec-make-disk-image}

`<nixpkgs/nixos/lib/make-disk-image.nix>` is a function to create _disk image_ in multiple formats: raw, QCOW2 (QEMU image format), QCOW2-Compressed (compressed version), VDI (Virtual Box native format), VPC (VirtualPC).

There are two operations mode for this function:

- either, you want only a Nix store disk image, it can be done using `cptofs` without any virtual machine involved ;
- or, you want a full NixOS install on the disk, it will start a virtual machine to perform various tasks such as partitionning, installation and bootloader installation.

When NixOS tests are running, this function gets called to generate potentially two kinds of disk images:

> - a Nix-store only disk image: useful when the test _do not need_ bootloader

Reviewer: It would be nice if you could go into some more detail, explaining what a Nix Store only image entails.

Author: I don't have good ideas to "extend it" extensively and suggest improving it incrementally. Feel free to suggest explanations :)

- a full-fledged NixOS installation disk image: useful when the test _do need_ to test the bootloader

## Features

> - whether to produce a Nix-store only image or not, **this can be incompatible with other options**

Author: As Nix Store only has non-features, I did a common / full NixOS sections.

### Image conversion phase

Using `qemu-img`, the disk image is converted from a raw format to the desired format: qcow2(-compressed), vdi, vpc.

### Partitionning script based on layouts

#### `none`

No partition table layout is written.

#### `legacy`

This partition table type is composed of a MBR and one primary ext4 partition starting at 1MiB extending to the full disk image.

It is unsuitable for UEFI.

#### `legacy+gpt`

This partition table type uses GPT and:

- create a "no filesystem" partition from 1MiB to 2MiB ;
- set `bios_grub` flag on this "no filesystem" partition, which marks it as a [GRUB BIOS partition](https://www.gnu.org/software/parted/manual/html_node/set.html) ;
- create a primary ext4 partition starting at 2MiB and extending to the full disk image ;
- perform optimal alignments checks on each partition

This partition has no ESP partition, it is unsuitable for EFI boot which requires an ESP partition, but it can work with CSM (Compatibility Support Module) which emulates BIOS for UEFI.

#### `efi`

This partition table type uses GPT and:

- creates an FAT32 ESP partition from 8MiB to specified `bootSize` parameter (256MiB by default), set it bootable ;

> - create a primary ext4 partition starting after the boot one and extending to the full disk image

Reviewer: ```suggestion
- creates an ext4 partition starting after the boot partition and extending to the full disk image
```

Author: I feel like it's important to keep `primary` as this is a technical explanation and the partition is neither extended neither logical.

#### `hybrid`

This partition table type uses GPT and:

- creates a "no filesystem" partition from 0 to 1MiB, set `bios_grub` flag on it ;
- creates an FAT32 ESP partition from 8MiB to specified `bootSize` parameter (256MiB by default), set it bootable ;
- creates a primary ext4 partition starting after the boot one and extending to the full disk image

This partition could be booted by a BIOS able to understand GPT layouts and recognizing the MBR at the start.

### How to run determinism analysis on results?

> Run your derivation with `--check` to rebuild it and verify it is the same.
> 
> Once it fails, you will be left with two folders with one having `.check`.

Reviewer: ```suggestion
Build your derivation with `--check` to rebuild it and verify it is the same.

If it fails, you will be left with two folders with one having `.check`.
```

> `diffoscope` is not able at the moment to diff two QCOW2 filesystems, it is advised to use raw format.
> 
> But even with raw formats, `diffoscope` cannot diff the partition table and partitions recursively.
> 
> For this, you can run `fdisk -l $image` and generate `dd if=$image of=$image-p$i.raw skip=$start count=$sectors` for each `(start, sectors)` listed in the `fdisk` output.
> 
> With this, you will have each partition as a separate file and you can diffoscope pairs of them to use `diffoscope` ability to read filesystems.

Reviewer: ```suggestion
You can use `diffoscope` to see the differences between the folders.
However, `diffoscope` is currently not able to diff two QCOW2 filesystems, thus, it is advised to use raw format.
Even if you use raw disks, `diffoscope` cannot diff the partition table and partitions recursively.
To solve this, you can run `fdisk -l $image` and generate `dd if=$image of=$image-p$i.raw skip=$start count=$sectors` for each `(start, sectors)` listed in the `fdisk` output. Now, you will have each partition as a separate file and you can compare them in pairs.
```
